synapse/tools/migrate_020.py

```diff
@@ -0,0 +1,1766 @@
+        '''
+        rprt = [
+            '\n',
+            f'Form counts for layer {iden}:',
+            f'{"FORM":<35}{"SRC_CNT":<15}{"DEST_CNT":<15}{"DIFF":<15}',
+        ]
+        src_tot = 0
+        dest_tot = 0
+        diff_tot = 0
+        for form in set(list(src_fcnt.keys()) + list(dest_fcnt.keys())):
+            scnt = src_fcnt.get(form, 0)
+            src_tot += scnt
+
+            dcnt = dest_fcnt.get(form, 0)
+            dest_tot += dcnt
+
+            diff = dcnt - scnt
+            diff_tot += diff
+
+            rprt.append(f'{form:<35}{scnt:<15}{dcnt:<15}{diff:<15}')
+            if addlog is not None:
+                await self._migrlogAdd(addlog, 'stat', f'{iden}:form:{form}', (scnt, dcnt))
+
+        rprt.append(f'{"TOTAL":<35}{src_tot:<15}{dest_tot:<15}{diff_tot:<15}')
+        rprt.append('\n')
+        prprt = '\n'.join(rprt)
+
+        return prprt
+
+    #############################################################
+    # Migration operations
+    #############################################################
+
+    async def _migrDirn(self):
+        '''
+        Setup the destination cortex dirn.  If dest already exists it will not be overwritten.
+        Copies all data *except* the layers
+
+        Returns:
+            (list): Discovered local physical layers
+        '''
+        migrop = 'dirn'
+
+        dest = self.dest
+        src = self.src
+        logger.info(f'Starting cortex dirn migration: {src} to {dest}')
+
+        lyrdir = os.path.join(src, 'layers')
+        locallyrs = []
+        for item in os.listdir(lyrdir):
+            if os.path.isdir(os.path.join(lyrdir, item)):
+                locallyrs.append(item)
+
+        logger.info(f'Found {len(locallyrs)} src physical layers.')
+        logger.debug(f'Source layers: {locallyrs}')
+
+        destexists = os.path.exists(dest)
+
+        if 'dirn' not in self.migrops:
+            logger.info(f'Skipping dirn migration step; dest exists={destexists}')
+            return locallyrs
+
+        if not destexists:
+            s_common.gendir(dest)
+
+        for sdir in os.listdir(src):
+            spath = os.path.join(src, sdir)
+            dpath = os.path.join(dest, sdir)
+
+            isdir = os.path.isdir(spath)
+            isfile = os.path.isfile(spath)
+            exists = os.path.exists(dpath)
+
+            if sdir == 'layers':
+                # make locallyr dirs if they don't exist but never overwrite
+                for lyr in locallyrs:
+                    lpath = os.path.join(dpath, lyr)
+                    if not os.path.exists(lpath):
+                        os.makedirs(lpath)
+                    else:
+                        logger.info(f'Layer dir exists, leaving as-is: {lyr}')
+
+            elif isfile:
+                if exists:
+                    os.remove(dpath)
+                shutil.copy(spath, dpath)
+
+            elif spath.endswith('slabs'):
```

Comment:
The axon directory has some slabs as well, it could use the backup copy too.

Reply:
good call; updated.